runtime/protocol/aws-protocol-core/common/src/aws/smithy/kotlin/runtime/awsprotocol/ClockSkewInterceptor.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.awsprotocol
+
+import aws.smithy.kotlin.runtime.ErrorMetadata
+import aws.smithy.kotlin.runtime.SdkBaseException
+import aws.smithy.kotlin.runtime.ServiceErrorMetadata
+import aws.smithy.kotlin.runtime.client.ProtocolRequestInterceptorContext
+import aws.smithy.kotlin.runtime.client.ResponseInterceptorContext
+import aws.smithy.kotlin.runtime.http.interceptors.HttpInterceptor
+import aws.smithy.kotlin.runtime.http.operation.HttpOperationContext
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.http.response.HttpResponse
+import aws.smithy.kotlin.runtime.http.response.header
+import aws.smithy.kotlin.runtime.telemetry.logging.logger
+import aws.smithy.kotlin.runtime.time.Instant
+import aws.smithy.kotlin.runtime.time.until
+import aws.smithy.kotlin.runtime.util.get
+import kotlinx.atomicfu.*
+import kotlin.coroutines.coroutineContext
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.minutes
+
+/**
+ * An interceptor used to detect clock skew (difference between client and server clocks) and apply a correction.
+ */
+public class ClockSkewInterceptor : HttpInterceptor {
+    public companion object {
+        /**
+         * How much must the clock be skewed before attempting correction
+         */
+        public val CLOCK_SKEW_THRESHOLD: Duration = 4.minutes
+
+        /**
+         * Determine whether the client's clock is skewed relative to the server.
+         * @return true if the service's response represents a definite clock skew error
+         * OR a *possible* clock skew error AND the skew exists. false otherwise.
+         * @param responseCodeDescription the server's response code description
+         * @param serverTime the server's time
+         */
+        internal fun Instant.isSkewed(serverTime: Instant, responseCodeDescription: String?): Boolean =
+            responseCodeDescription?.let {
+                CLOCK_SKEW_ERROR_CODES.contains(it) || (POSSIBLE_CLOCK_SKEW_ERROR_CODES.contains(it) && until(serverTime).absoluteValue >= CLOCK_SKEW_THRESHOLD)
+            } ?: false
+
+        // Errors definitely caused by clock skew
+        private val CLOCK_SKEW_ERROR_CODES = listOf(
+            "RequestTimeTooSkewed",
+            "RequestExpired",
+            "RequestInTheFuture",
+        )
+
+        // Errors possibly caused by clock skew
+        private val POSSIBLE_CLOCK_SKEW_ERROR_CODES = listOf(
+            "PriorRequestNotComplete",
+            "RequestTimeout",
+            "RequestTimeoutException",
+            "InternalError",
+        )
```

Comment:
**Correctness**: These don't appear to be related to clock skew. Looks like [Java considers `InvalidSignatureException`, `SignatureDoesNotMatch`, and `AuthFailure` as possible errors](https://github.com/aws/aws-sdk-java-v2/blob/586b471405bd504d33389c93a436d33702bd7d8e/core/aws-core/src/main/java/software/amazon/awssdk/awscore/internal/AwsErrorCode.java#L57-L61).

Reply:
🤦 Good catch. I took the wrong list of exceptions from Java and then convinced myself they could be caused by clock skew.